### ISSUE TITLE
Reinitialize the random generator in worker processes of `DataLoader`

### ISSUE THREAD

**Issue by netheril96** [pytorch/pytorch#3880]
`torch.utils.data.DataLoader` can use multiprocessing to load and preprocess the data. It is commonly used to overlap the GPU computation and data loading. It has a minor flaw, though. On Unix, the processes are by default created with `fork`, so the global random state is copied. If each worker does something pseudo-randomly (common in data augmentation), their actions are all the same. For example, random crop is commonly used in image classification task, and in this case, all `DataLoader` workers will crop image in the same way, reducing the effectiveness of data augmentation.

I propose a new option `worker_init_fn` be added to `DataLoader` constructor, and the function is called with the process ID right in the beginning of each worker loop. By default, the function ignores the process ID and reseeds both pytorch and numpy default random generator. The users can override it to do other things that each worker might need differently.

**Comment by soumith** [comment]
this seems like a reasonable solution,. though needs some thought on the interface. We cant rely on `set_start_method` all the time right. wdyt @colesbury ?

**Comment by colesbury** [comment]
Yes, this seems reasonable. @apaszke suggested using the worker ID (i.e. 0, 1, ... N-1) instead of PID to preserve determinism when the user sets the torch.manual_seed.

I think you probably want to combine the worker ID with the current RNG state so that changing the seed in the main process changes affects the seeds in the worker processes.

**Comment by vfdev-5** [comment]
I tried to reproduce this issue and there is an interesting behaviour (at least for me) that I observe. If random data augmentation is based on `np.random` then forking the dataset with random transformations produces batches with exactly the same augmentations. However, if random data augmentation is based on `random` from python stdlib there is no problem with data augmentations between batches. [Here](https://gist.github.com/vfdev-5/aa61ea2bac2494ca852963fc941fcd7e) is gist

**Comment by SsnL** [comment]
@vfdev-5 that is because np.random doesn't work nicely with multiprocessing without manual seeding.

**Comment by SsnL** [comment]
looking into this.
